### PR TITLE
Add functionality to use custom image names for wave builds

### PR DESCRIPTION
**Pull request by munishchouhan**
This PR will add the functionality to use imageName provided by the user to create targetImage for wave build container.

This PR is dependent on 
https://github.com/seqeralabs/libseqera/pull/13

### CODE REVIEW

**Comment by pditommaso** [comment]
Please let's try to keep the title in one line and surely not exceeding 2 lines

File: src/main/groovy/io/seqera/wave/controller/ContainerTokenController.groovy [hunk 1]
```diff
@@ -222,6 +223,10 @@ class ContainerTokenController {
             throw new BadRequestException("Missing build repository attribute")
         if( !buildConfig.defaultCacheRepository )
             throw new BadRequestException("Missing build cache repository attribute")
+        //validate custom image name
+        if( req.imageName && !RegHelper.isValidImageName(req.imageName) ){
+            throw new BadRequestException("Supplied container image name '${req.imageName}' is not validate")
```
Comment (marcodelapierre) [inline src/main/groovy/io/seqera/wave/controller/ContainerTokenController.groovy:228]: shouldn't it end with `is not valid` ?

File: src/test/groovy/io/seqera/wave/util/RegHelperTest.groovy [hunk 2]
```diff
@@ -284,4 +283,23 @@ class RegHelperTest extends Specification {
...
+
+    @Unroll
+    def 'should validate the provided custom container image name'(){
+        expect:
+        RegHelper.isValidImageName( IMAGENAME ) == ISVALID
+
+        where:
+        IMAGENAME               | ISVALID
+        'foo'                   | true
+        'foo:1.1'               | true
```
Comment (marcodelapierre) [inline src/test/groovy/io/seqera/wave/util/RegHelperTest.groovy:295]: The colon symbol should not be allowed, right? 

This is because when we provide an `imageName`, the image tag is still defined by Wave (see e.g. lines 197-204 in your edits to ContainerTokenController).

File: src/main/groovy/io/seqera/wave/service/builder/BuildRequest.groovy [hunk 3]
```diff
@@ -149,7 +149,7 @@ class BuildRequest {
         this.buildContext = buildContext
         this.condaFile = condaFile
         this.spackFile = spackFile
-        this.targetImage = makeTarget(format, repo, containerId, condaFile, spackFile)
+        this.targetImage = makeTarget(format, repo, containerId, condaFile, spackFile, imageName)
```
Comment (marcodelapierre) [inline src/main/groovy/io/seqera/wave/service/builder/BuildRequest.groovy:152]: I am missing something here.  Where does `imageName` come from?

I would have expected it to come as a new last argument of this `BuildRequest` initialiser method (line 145 above), considering  call to this method has been modified accordingly in `ContainerTokenController` lines 243-259.

**Review by marcodelapierre — CHANGES_REQUESTED** [review]

File: src/main/groovy/io/seqera/wave/service/builder/BuildRequest.groovy (hunk 3 above)
Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/service/builder/BuildRequest.groovy:156]: yes sorry, it was deleted while resolving the conflicts added back

File: src/test/groovy/io/seqera/wave/util/RegHelperTest.groovy (hunk 2 above)
Comment (munishchouhan) [reply src/test/groovy/io/seqera/wave/util/RegHelperTest.groovy]: done

File: src/main/groovy/io/seqera/wave/controller/ContainerTokenController.groovy (hunk 1 above)
Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/controller/ContainerTokenController.groovy]: done

**Comment by marcodelapierre** [comment]
What is the rationale behind renaming `ContainerTokenController` to `ContainerController` ?

File: settings.gradle [hunk 4]
```diff
@@ -3,4 +3,4 @@ rootProject.name="wave"
 // only for development
 // clone https://github.com/seqeralabs/libseqera
 // in a sibling directory and include the build below
-//includeBuild('../libseqera')
+includeBuild('../libseqera')
```
Comment (marcodelapierre) [inline settings.gradle:6]: this should not enter the PR

File: src/test/groovy/io/seqera/wave/util/RegHelperTest.groovy [hunk 5]
```diff
@@ -284,4 +283,23 @@ class RegHelperTest extends Specification {
...
+    @Unroll
+    def 'should validate the provided custom container image name'(){
+        expect:
+        RegHelper.isValidImageName( IMAGENAME ) == ISVALID
+
+        where:
+        IMAGENAME               | ISVALID
+        'foo'                   | true
+        'foo:'                  | false
+        'foo:SHA123asdf'        | false
```
Comment (marcodelapierre) [inline src/test/groovy/io/seqera/wave/util/RegHelperTest.groovy:296]: I would edit these 2 to explicitly check about sha256 reference syntax:
```
'foo:SHA123asdf' -> 'foo@sha256:asdf'

'foo/bar/foo:SHA123asdf' -> 'foo/bar/foo@sha256:asdf'
cowsay@sha256:d24d
```

File: src/main/resources/application.yml [hunk 6]
```diff
@@ -64,8 +64,8 @@ wave:
     kaniko-image: "gcr.io/kaniko-project/executor:v1.19.2"
     singularity-image: "quay.io/singularity/singularity:v3.11.4-slim"
     singularity-image-arm64: "quay.io/singularity/singularity:v3.11.4-slim-arm64"
-    repo: "195996028523.dkr.ecr.eu-west-1.amazonaws.com/wave/build/dev"
-    cache: "195996028523.dkr.ecr.eu-west-1.amazonaws.com/wave/build/cache"
+    repo : "195996028523.dkr.ecr.eu-west-1.amazonaws.com/wave/build/dev"
```
Comment (marcodelapierre) [inline src/main/resources/application.yml:67]: are these changes needed?

File: src/main/groovy/io/seqera/wave/util/RegHelper.groovy [hunk 7]
```diff
@@ -292,4 +292,9 @@ class RegHelper {
             log.debug "Unexpected error while closing http response - cause: ${e.message}", e
         }
     }
+
+    static boolean isValidImageName(String imageName){
+        final DOCKER_IMAGE_REGEX = /^(?:[a-zA-Z0-9-]+\.)?[a-zA-Z0-9-]+(?:\/[a-zA-Z0-9-]+)*$/
```
Comment (marcodelapierre) [inline src/main/groovy/io/seqera/wave/util/RegHelper.groovy:297]: Last one, see this description around separators characters within components of the image name:

```
The path consists of slash-separated components. Each component may contain lowercase letters, digits and separators. A separator is defined as a period, one or two underscores, or one or more hyphens. A component may not start or end with a separator.
```
can you make sure this is accounted for?

References:
- [Docker docs](https://docs.docker.com/reference/cli/docker/image/tag/#description)
- [OCI Distribution Spec](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pulling-manifests)

**Review by marcodelapierre — CHANGES_REQUESTED** [review]
Some last minor requests

File: settings.gradle (hunk 4 above)
Comment (munishchouhan) [reply settings.gradle]: done

File: src/main/resources/application.yml (hunk 6 above)
Comment (munishchouhan) [reply src/main/resources/application.yml]: no, reverted

File: src/main/groovy/io/seqera/wave/util/RegHelper.groovy (hunk 7 above)
Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/util/RegHelper.groovy:297]: to make it par with OCI specification, I am using regex from https://github.com/opencontainers/distribution-spec/blob/main/spec.md#pulling-manifests

Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/util/RegHelper.groovy:297]: so it should be good now

File: src/test/groovy/io/seqera/wave/util/RegHelperTest.groovy (hunk 5 above)
Comment (munishchouhan) [reply src/test/groovy/io/seqera/wave/util/RegHelperTest.groovy]: done

File: src/main/groovy/io/seqera/wave/util/RegHelper.groovy (hunk 7 above)
Comment (marcodelapierre) [reply src/main/groovy/io/seqera/wave/util/RegHelper.groovy]: thank you 👍

**Review by marcodelapierre — APPROVED** [review]

**Comment by marcodelapierre** [comment]
Note: successful CI requires merging and release of Libseqera - https://github.com/seqeralabs/libseqera/pull/13

**Comment by marcodelapierre** [comment]
@pditommaso  @munishchouhan  I would use the custom image name to compute the container digest, as the build repository name is also used.   What do you think?

See commit https://github.com/seqeralabs/wave/pull/408/commits/1651e7046a37a7f2c90a2b4c1b24e8a19ab85e6c

**Comment by marcodelapierre** [comment]
Ready for approval and merge @pditommaso

**Comment by munishchouhan** [comment]
> It would be useful to add unit tests that validate the method `BuildRequest.makeTarget` as described in the requirement [here](https://github.com/seqeralabs/wave/issues/395)

@pditommaso already added here 
https://github.com/seqeralabs/wave/blob/f90d7db026d92cda6a6396ef57afe424b1be1f47/src/test/groovy/io/seqera/wave/service/builder/BuildRequestTest.groovy#L269-L278

**Comment by pditommaso** [comment]
Can it be refactored as a `where:` table to make it more readable ?

**Comment by munishchouhan** [comment]
> Can it be refactored as a `where:` table to make it more readable ?

done

**Comment by marcodelapierre** [comment]
@munishchouhan just a reminder, from the meeting, to document the custom image name for the API endpoint - thank you!

**Comment by munishchouhan** [comment]
> @munishchouhan just a reminder, from the meeting, to document the custom image name for the API endpoint - thank you!

done

File: src/test/groovy/io/seqera/wave/service/builder/BuildRequestTest.groovy [hunk 8]
```diff
@@ -275,7 +275,10 @@ class BuildRequestTest extends Specification {
         'quay.io/org/name'      | ''                  | 'quay.io/org/name:12345'
         'quay.io/org/name'      | 'foo'               | 'quay.io/org/name/foo:12345'
         'quay.io/org'           | 'foo/bar'           | 'quay.io/org/foo/bar:12345'
+        'quay.io/org/'          | 'foo/bar'           | 'quay.io/org/foo/bar:12345'
         'quay.io'               | 'foo/bar'           | 'quay.io/foo/bar:12345'
+        'quay.io/'              | 'foo/bar'           | 'quay.io/foo/bar:12345'
+        'quay.io/'              | 'foo/bar'           | 'quay.io/foo/bar:12345'
```
Comment (munishchouhan) [inline src/test/groovy/io/seqera/wave/service/builder/BuildRequestTest.groovy:281]: same line twice. we can remove 281

Comment (pditommaso) [reply src/test/groovy/io/seqera/wave/service/builder/BuildRequestTest.groovy]: Right

**Comment by pditommaso** [comment]
Closing in favour of #460
